### PR TITLE
Remove MlsRules

### PR DESCRIPTION
**Pull request by mulmarta**
This looks quite big but most changes are pretty trivial.

The real change is to
 * replace commit_options by appropriate functions of CommitBuilder
 * introduce ProposalBuilder
 * replace encryption_options by appropriate functions of CommitBuilder and ProposalBuilder

## The only relevant files are
 * group/mod.rs  (new API)
 * group/proposal.rs  (ProposalBuilder)
 * group/commit/builder.rs  (new options for CommitBuilder)

## Follow-up
 * add any missing functionality to external group
 * add branch / reinit builders

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT license.

### CODE REVIEW

File: mls-rs/src/group/mod.rs [hunk 1]
```diff
@@ -995,6 +966,23 @@ where
...
+            authenticated_data,
+            PaddingMode::None,
+        )
+        .await
+    }
+
+    // TODO make a builder
+    #[cfg(feature = "private_message")]
+    #[cfg_attr(not(mls_build_async), maybe_async::must_be_sync)]
+    pub async fn encrypt_application_message_with_padding(
```
Comment (tomleavy) [inline mls-rs/src/group/mod.rs:981]: I feel like in the future even stuff like this can be handled via the default above and some builder that deals with authenticated_data as well

**Review by tomleavy — DISMISSED** [review]
Looks good, see nit comments. One thing we should keep in mind as we get closer to where we want to be with the API is that we need to reorg the tests at some point to be able to be at the properly level. We should have builder level tests for edge cases and then integration tests for just general protocol testing. That will make some of these files a lot smaller

**Comment by tomleavy** [comment]
Any idea why the tests above are failing? Seems like something new?

**Comment by mulmarta** [comment]
> Any idea why the tests above are failing? Seems like something new?

It's new but unrelated, they fail on main too. We should investigate separately.

File: mls-rs/src/group/mod.rs (hunk 1 above)
Comment (mulmarta) [reply mls-rs/src/group/mod.rs:981]: Agreed, that's what the TODO comment is about.

File: mls-rs/src/group/proposal.rs [hunk 2]
```diff
@@ -589,3 +589,298 @@ impl From<ProposalRef> for ProposalOrRef {
...
+        Update,
+        #[cfg(feature = "psk")]
+        ExternalPsk {
+            psk_id: ExternalPskId,
+        },
+        #[cfg(feature = "psk")]
+        ResumptionPsk {
+            psk_id: ResumptionPsk,
+        },
+        ReInit,
```
[the remark this answers is not in the document]
Comment (mulmarta) [reply mls-rs/src/group/proposal.rs:660]: This way you won't be able to make a new group_id in the builder?

[the remark this answers is not in the document]
Comment (tomleavy) [reply mls-rs/src/group/proposal.rs:660]: Ok yeah, just thought it was weird that reinit specific options are in a flat set at the top level and not somehow bundled together. It honestly doesn't matter much.

**Review by tomleavy — APPROVED** [review]
